### DISCUSSION TITLE
Use `take` (`LIMIT`) in compiled queries

### DISCUSSION THREAD

**Discussion by sebarys** [slick/slick#3055] [Questions]
I wanted to limit number of returned rows by some function argument:
```
private lazy val myCompiledQuery = Compiled {
    (
      namePattern: Rep[String],
      limit: Rep[Long],
    ) =>
      queries
        .getUsersThatStartsWith(namePattern)
        .take(limit)
  }
```
but it don't work unfortunately
```
overloaded method take with alternatives:
  (num: Int)slick.lifted.Query[A,B,Seq] <and>
  (num: Long)slick.lifted.Query[A,B,Seq] <and>
  (num: slick.lifted.ConstColumn[Long])slick.lifted.Query[A,B,Seq]
 cannot be applied to (slick.lifted.Rep[Long])
        .take(limit)
```
is there any way to parameterize LIMIT in compiled query?

**Comment by nafg** [comment]
Can Compiled query take a raw (non Rep) parameter?

**Reply by sebarys** [reply]
I was thinking about it, but it fail during compilation:
```
  lazy val compiledQuery = Compiled { (userId: Rep[UUID], limit: Long) =>
    Tables.Users.filter(_.id === userId).take(limit)
  }

Computation of type (slick.lifted.Rep[UUID], Long) => slick.lifted.Query[Tables.Users,Tables.Users#TableElementType,Seq] cannot be compiled (as type C)
  lazy val compiledQuery = Compiled { (userId: Rep[UUID], limit: Long) =>

```

**Reply by sebarys** [reply]
Compiled query has `.extract` method so I'm currently converting `CompiledQuery` back to `Query` and apply `.take(...)`, but I'm not sure whether it isn't defeat the purpose of compiled queries (generate SQL once, not on each invocation).
```
val baselineQuery = compiledQuery.apply(userId)
baselineQuery.extract.take(limit).result
```

I'm not sure how it can be validated

**Reply by nafg** [reply]
Not 100% sure but I think it does defeat the purpose.

**Comment by nafg** [comment]
I guess the right solution would be to overload `take` to accept a Rep, do you think you can take a stab?

**Reply by sebarys** [reply]
I will look into it next week

**Comment by sebarys** [comment]
@nafg 
I see that take has variant with `ConstColumn` 
```
  /** Select the first `num` elements. */
  def take(num: ConstColumn[Long]): Query[E, U, C] = new WrappingQuery[E, U, C](Take(toNode, num.toNode), shaped)
  /** Select the first `num` elements. */
  def take(num: Long): Query[E, U, C] = take(LiteralColumn(num))
  /** Select the first `num` elements. */
  def take(num: Int): Query[E, U, C] = take(num.toLong)
```
and `ConstColumn` extends Rep: `class ConstColumn[T : TypedType](val toNode: Node) extends Rep.TypedRep[T] {`
I've checked and it is possible to use `ConstColumn` instead of `Rep` as Compiled query param

query definition:
```
  lazy val compiledQuery = Compiled { (userId: Rep[UUID], limit: ConstColumn[Long]) =>
    Tables.Users.filter(_.id === userId).take(limit)
  }
```

usage:
```
val userId: UUID = ...
val limit: Long = ... 
compiledQuery.apply((userId, limit))
```

now, I'm thinking whether Compiled query shouldn't require `ConstColumn` (or `LiteralColumn`) instead of `Rep`, as whole idea is that you want pass to compiled query runtime known constants 🤔
